File: solutions/code-carl/ch01_array/T5_LC0209_MinSubArrayLen.java
/**
 * Practice template for:
 * - LeetCode 209. Minimum Size Subarray Sum
 *   https://leetcode.cn/problems/minimum-size-subarray-sum/
 */
public class T5_LC0209_MinSubArrayLen {
    public int minSubArrayLen(int target, int[] nums) {
        int sum = 0, j = 0, minLen = Integer.MAX_VALUE;
        for (int i  = 0; i < nums.length; i++) {
            sum += nums[i];
            while (sum >= target) {
                minLen = Math.min(minLen, i - j + 1);
                sum -= nums[j];
                j++;
            }
        }
        return minLen == Integer.MAX_VALUE ? 0 : minLen;
    }
}

File: solutions/python/code-carl/ch01_array/t5_lc0209_min_sub_array_len.py
"""LeetCode 209. Minimum Size Subarray Sum
https://leetcode.cn/problems/minimum-size-subarray-sum/
"""
import sys
from typing import List


class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        i, currSum, minLen = 0, 0, sys.maxsize
        for j, _ in enumerate(nums):
            currSum += nums[j]
            while currSum >= target:
                currSum -= nums[i]
                minLen = min(minLen, j - i + 1)
                i += 1
        return 0 if minLen == sys.maxsize else minLen


if __name__ == "__main__":
    solution = Solution()
    assert solution.minSubArrayLen(7, [2, 3, 1, 2, 4, 3]) == 2
    assert solution.minSubArrayLen(4, [1, 4, 4]) == 1
    assert solution.minSubArrayLen(11, [1, 1, 1, 1, 1, 1, 1, 1]) == 0
    print("All tests passed.")
